src/ApacheOrcDotNet.OptimizedReaderTest.App/TradeDataSource.cs

```diff
@@ -8,120 +8,182 @@
+            _lookup = lookup;
+            _columns = columns;
+            _buffers = buffers;
+            _matchCandidates = matchCandidates;
 
-            foreach (var stripeId in stripeIds)
-            {
-                var rowGroupIndexes = _reader.FilterRowGroups(stripeId, sourceColumn, sourceFilterValues);
-                rowGroupIndexes = _reader.FilterRowGroups(rowGroupIndexes, stripeId, symbolColumn, symbolFilterValues);
-                rowGroupIndexes = _reader.FilterRowGroups(rowGroupIndexes, stripeId, timeColumn, timeFilterValues);
-
-                _filters.Add(stripeId, rowGroupIndexes.ToList());
-            }
+            ApplyFilters();
         }
 
-        public int ApproxRowCount
+        public int ApproxRowCount => _matches.Count * _reader.MaxValuesPerRowGroup;
+
+        public int ReadBatch(int writeStartIndex, Span<decimal?> times, Span<decimal?> prices, Span<long?> sizes)
         {
-            get
+            var writeIndex = 0;
+            var startTimeTotalSeconds = (decimal)_lookup.StartTime.TotalSeconds;
+            var endTimeTotalSeconds = (decimal)_lookup.EndTime.TotalSeconds;
+
+            while (_matches.Count > 0)
             {
-                var rowGroups = _filters.Values.Sum(v => v.Count);
-                return rowGroups * _reader.MaxValuesPerRowGroup;
+                var (stripeId, rowEntryIndex) = _matches.Dequeue();
+
+                Task.WhenAll(
+                    _reader.LoadDataAsync(stripeId, rowEntryIndex, _buffers.Source),
+                    _reader.LoadDataAsync(stripeId, rowEntryIndex, _buffers.Symbol),
+                    _reader.LoadDataAsync(stripeId, rowEntryIndex, _buffers.Time),
+                    _reader.LoadDataAsync(stripeId, rowEntryIndex, _buffers.Size),
+                    _reader.LoadDataAsync(stripeId, rowEntryIndex, _buffers.Price)
+                ).Wait();
+
+                for (int idx = 0; idx < _reader.NumValuesLoaded; idx++)
+                {
+                    var source = _buffers.Source.Values[idx];
+                    var symbol = _buffers.Symbol.Values[idx];
+                    var time = _buffers.Time.Values[idx];
+
+                    if (source == _lookup.Source && symbol == _lookup.Symbol && time >= startTimeTotalSeconds && time <= endTimeTotalSeconds)
+                    {
+                        _valuesFound = true;
+
+                        times[writeStartIndex + writeIndex] = _buffers.Time.Values[idx];
+                        sizes[writeStartIndex + writeIndex] = _buffers.Size.Values[idx];
+                        prices[writeStartIndex + writeIndex] = _buffers.Price.Values[idx];
+
+                        writeIndex++;
+                    }
+                }
+
+                // We only keep looking for values if
+                // nothing has been found before.
+                if (writeIndex == 0 && !_valuesFound)
+                    continue;
+
+                break;
```

Comment:
I get why this is here but it feels like an antipattern.  `break` should be the exception, not the rule.  `continue` should be the rule, not the exception.